alf/utils/video_recorder.py

```diff
@@ -179,23 +221,199 @@ def capture_frame(self, pred_info=None):
+            # not enough future date for defer encoding
+            return
+
+        if rewards.ndim > 1:
+            # slice the multi-dimensional rewards
+            # assume the first dimension is the overall reward
+            rewards = rewards[..., 0]
+
+        actions = torch.cat(self._action_buffer, dim=0)
+
+        num_steps = self._future_steps + 1
+
+        reward_curve_set = []
+        action_curve_set = []
+        predictive_curve_set = []
+
+        encoding_steps = episode_buffer_length if encode_all else 1
+        for t in range(encoding_steps):
+            H = min(num_steps,
+                    episode_buffer_length - t)  # total display steps
+            if H > 0:
+                t_actions = actions[t:t + H]
+                t_rewards = rewards[t:t + H]
+
+                if info_func is not None:
+                    predictions = info_func(self._observation_buffer[t],
+                                            t_actions)
+                    assert predictions.ndim == 1 or predictions.shape[1] == 1, \
+                        "only support displaying scalar predictive information"
+                    predictions = predictions.view(-1).detach().cpu().numpy()
+
+                    pred_curve = self._plot_value_curve(
+                        "prediction", [predictions],
+                        legends=["Prediction"],
+                        fig_size=6,
+                        height=300,
+                        width=300,
+                        linewidth=5)
+                    predictive_curve_set.append(pred_curve)
+
+                reward_gt = t_rewards.view(-1).cpu().numpy()
+                action_cpu = t_actions.detach().cpu().numpy()
+
+                reward_curve = self._plot_value_curve(
+                    "rewards", [reward_gt],
+                    legends=["GroundTruth"],
+                    fig_size=6,
+                    height=300,
+                    width=300,
+                    linewidth=5)
+                reward_curve_set.append(reward_curve)
+
+                action_curve = self._plot_value_curve(
+                    "actions",
+                    [action_cpu[..., i] for i in range(action_cpu.shape[-1])],
+                    legends=[
+                        "a" + str(i) for i in range(action_cpu.shape[-1])
+                    ],
+                    fig_size=6,
+                    height=300,
+                    width=300,
+                    linewidth=5)
+                action_curve_set.append(action_curve)
+
+            self._observation_buffer.pop(0)
+            self._reward_buffer.pop(0)
+            self._action_buffer.pop(0)
+
+        # encode all frames
+        self._encode_frames_in_buffer_with_external(
+            [reward_curve_set, action_curve_set, predictive_curve_set])
+
+    def _encode_frames_in_buffer_with_external(self, set_of_external_frames):
+        """ Encode jointly internal and external frames
```

Comment:
The word "internal" and "external" are unintuitive. Better choices should be considered.

Reply:
Updated with more descriptive names